### PR TITLE
Improve detection of cgroups v2 support for memory controllers

### LINKED ISSUE #62788: cgroups v2 detection fails in containerized environment
This is a follow-up issue based on late discussion in #59435

**Describe the unexpected behaviour**
New cgroups v2 detection introduced in ad13a57b doesn't work when clickouse runs in container. 

**How to reproduce**
* ClickHouse local version 24.3.2.23 (official build).
* Start in docker on a system with cgroups v2 and memory limit set.
```
docker run -it --rm --ulimit nofile=262144:262144  -m 2g clickhouse/clickhouse-server:24.3.2.23-alpine
```

**Expected behavior**
`CgroupsMemoryUsageObserver` properly detects cgroups v2 and sets max memory based on `memory.max` cgroup setting.

**Error message and/or stacktrace**
```
2024.04.19 08:46:53.884176 [ 42 ] {} <Error> Application: Disabling cgroup memory observer because of an error during initialization: Code: 107. DB::Exception: Cannot find cgroups v1 or v2 current memory file. (FILE_DOESNT_EXIST), Stack trace (when copying this message, always include the lines below):

0. DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000cbcedbb
1. DB::Exception::Exception<>(int, FormatStringHelperImpl<>) @ 0x0000000007668f23
2. DB::CgroupsMemoryUsageObserver::CgroupsMemoryUsageObserver(std::chrono::duration<long long, std::ratio<1l, 1l>>) @ 0x000000000cceadce
3. DB::Server::main(std::vector<String, std::allocator<String>> const&) @ 0x000000000cd5cb5d
...
```

**Additional context**
The problem is in detection of enabled controllers for cgroup hierarchies, when it is checked within container. The `/sys/fs/cgroup/cgroup.subtree_control` is empty and therefore the check `cgroupsV2MemoryControllerEnabled()` returns false.

Example:
```
# on physical host:
$ head /sys/fs/cgroup/cgroup.controllers /sys/fs/cgroup/cgroup.subtree_control
==> /sys/fs/cgroup/cgroup.controllers <==
cpuset cpu io memory hugetlb pids rdma misc

==> /sys/fs/cgroup/cgroup.subtree_control <==
cpuset cpu io memory hugetlb pids rdma misc

# in container:
docker run   --rm -m 100m alpine:3.10 head /sys/fs/cgroup/cgroup.controllers /sys/fs/cgroup/cgroup.subtree_control
==> /sys/fs/cgroup/cgroup.controllers <==
cpuset cpu io memory hugetlb pids rdma misc

==> /sys/fs/cgroup/cgroup.subtree_control <==

```

cc: @rschu1ze @Stanislav-Vlasov

### PR DESCRIPTION
**Pull request by rschu1ze**
Fixes #62788

Related to https://github.com/ClickHouse/ClickHouse/issues/62788#issuecomment-2074538377 and an earlier try to fix the issue: https://github.com/ClickHouse/ClickHouse/pull/62900

CGroups is a Linux subsystem for allocating resources (e.g. CPU, memory, I/O) to processes. For that, the administrator defines a tree of cgroups in `/sys/fs/cgroups`, and assigns processes to specific nodes in this tree. Each resource belongs to a CGroup "controller", e.g. the memory controller.

CGroups also provides a means for processes to get useful information about the current and maximum resource consumption. Different places in ClickHouse read such information from cgroups. In order to do that, they need to identify the cgroup of the ClickHouse server process, but also perform sanity checks like "Is the memory controller in CGroups even enabled?" (if no --> skip read or throw).

To detect if cgroups v2 has its "memory" controller enabled, ClickHouse previously read the top-level `cgroup.subtree_control` file, equally named files in intermediate or leaf nodes were ignored.

After reading more cgroups docs, I think the previous code was wrong. The problem is that `cgroup.subtree_control` controls the children of the current cgroup but not the current cgroup itself. As a result, one could have a perfectly legal root cgroup without child nodes with
- `cgroup.controllers` containing `[...] memory [...]` and
- `cgroup.subtree_controllers` being empty (because there are no child cgroups anyways).

This happened in issue #62788 (a containerized environment) and ClickHouse would think that the system has no "memory" controller.

Most non-container environments have recursive cgroups defined in which `cgroups.subtree_controllers` contains the same values as `cgroups.controllers`, so it worked for most users most of the time.

With this PR, ClickHouse now instead reads the bottom-most nested `cgroup.controllers` file (corresponding to the processe's cgroup). Based on the docs, I am pretty sure that this file always exists and we don't need to read the same file in parent nodes.

### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Improve the detection of cgroups v2 memory controller in unusual locations. This fixes a warning that the cgroup memory observer was disabled because no cgroups v1 or v2 current memory file could be found.

### CODE REVIEW

**Comment by robot-ch-test-poll3** [comment]
*This is an automated comment for commit a826509a860cce86e9cd19d12af700f7fb4feb0d with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/62903/a826509a860cce86e9cd19d12af700f7fb4feb0d/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

**Comment by rschu1ze** [comment]
@jkartseva I updated the PR description (hope it is now more clear). Like to check / approve? Thanks!

**Comment by jkartseva** [comment]
Is the fix for detecting available controllers or enabled controllers?

https://docs.kernel.org/admin-guide/cgroup-v2.html:

> No controller is enabled by default.  Controllers can be enabled and
> disabled by writing to the "cgroup.subtree_control" file::
> 
>   ```# echo "+cpu +memory -io" > cgroup.subtree_control```

Also, from the docs:
> Enabling a controller in a cgroup indicates that the distribution of the target resource across its immediate children will be controlled. Consider the following sub-hierarchy. The enabled controllers are listed in parentheses:
> 
> ```
> A(cpu,memory) - B(memory) - C()
>                           \ D()
> ```

> As A has “cpu” and “memory” enabled, A will control the distribution of CPU cycles and memory to its children, in this case, B. As B has “memory” enabled but not “CPU”, C and D will compete freely on CPU cycles but their division of memory available to B will be controlled.


Wouldn't the fix be checking the `cgroup.subtree_control` of both the immediate parent can the target cgroup?

File: base/base/cgroupsv2.cpp [hunk 1]
```diff
@@ -23,18 +23,19 @@ bool cgroupsV2MemoryControllerEnabled()
...
         return false;
     std::string controllers;
-    std::getline(subtree_control_file, controllers);
-    if (controllers.find("memory") == std::string::npos)
-        return false;
-    return true;
+    std::getline(controllers_file, controllers);
+    if (controllers.find("memory") != std::string::npos)
+        return true;
+    return false;
```
Comment (jkartseva) [inline base/base/cgroupsv2.cpp:38]: nit: `return controllers.find("memory") != std::string::npos;`

File: base/base/cgroupsv2.cpp [hunk 2]
```diff
@@ -23,18 +23,19 @@ bool cgroupsV2MemoryControllerEnabled()
...
     chassert(cgroupsV2Enabled());
-    /// According to https://docs.kernel.org/admin-guide/cgroup-v2.html:
-    /// - file 'cgroup.controllers' defines which controllers *can* be enabled
-    /// - file 'cgroup.subtree_control' defines which controllers *are* enabled
-    /// Caveat: nested groups may disable controllers. For simplicity, check only the top-level group.
-    std::ifstream subtree_control_file(default_cgroups_mount / "cgroup.subtree_control");
-    if (!subtree_control_file.is_open())
+    /// According to https://docs.kernel.org/admin-guide/cgroup-v2.html, file "cgroup.controllers" defines
+    /// which controllers are enabled for the current cgroup. Check the bottom-most nested "cgroup.controllers"
+    /// file.
```
Comment (jkartseva) [inline base/base/cgroupsv2.cpp:28]: https://docs.kernel.org/admin-guide/cgroup-v2.html says differently: `cgroup.controllers` is for available controllers.
It is worth extending the comment.

**Review by jkartseva — APPROVED** [review]
Please take a look at the comments.

**Comment by rschu1ze** [comment]
@jkartseva Thanks, I'll adress your comments here: #63260

**Pull request merged**